node/pkg/governor/governor.go

```diff
@@ -93,19 +94,85 @@ type (
 		dbData db.PendingTransfer // This info gets persisted in the DB.
 	}
 
-	// Payload of the map of chains being monitored
+	// Used in flow cancel calculations. Wraps a database Transfer. Also contains a signed amount field in order to
+	// hold negative values. This field will be used in flow cancel calculations to reduce the Governor usage for a
+	// supported token.
+	transfer struct {
+		dbTransfer *db.Transfer
+		value      int64
+	}
+
+	// Payload of the map of chains being monitored. Contains transfer data for both emitted and received transfers.
+	// `transfers` with positive Value represent outgoing transfers from the emitterChainId. Transfers with negative
+	// Value represent incoming transfers of Assets that can Flow Cancel.
 	chainEntry struct {
 		emitterChainId          vaa.ChainID
 		emitterAddr             vaa.Address
 		dailyLimit              uint64
 		bigTransactionSize      uint64
 		checkForBigTransactions bool
 
-		transfers []*db.Transfer
+		transfers []transfer
 		pending   []*pendingEntry
 	}
 )
 
+// newTransferFromDbTransfer function    Perform a bounds check on dbTransfer.Value to ensure it can fit into int64.
+// This should always be the case for normal operation as dbTransfer.Value represents the USD value of a transfer.
+func newTransferFromDbTransfer(dbTransfer *db.Transfer) (tx transfer, err error) {
+	if dbTransfer.Value > math.MaxInt64 {
+		return tx, fmt.Errorf("Value for db.Transfer exceeds MaxInt64: %d", dbTransfer.Value)
+	}
+	return transfer{dbTransfer, int64(dbTransfer.Value)}, nil
+}
+
+// addFlowCancelTransfer method    Appends a transfer to a ChainEntry's transfers property.
+// SECURITY: This method performs validation to ensure that the Flow Cancel transfer is valid. This is important to
+// ensure that the Governor usage cannot be lowered due to malicious or invalid transfers.
+// - the Value must be negative (in order to represent an incoming value)
+// - the TargetChain must match the chain ID of the Chain Entry
+func (ce *chainEntry) addFlowCancelTransfer(transfer transfer) error {
+	value := transfer.value
+	targetChain := transfer.dbTransfer.TargetChain
+	if value > 0 {
+		return fmt.Errorf("Flow cancel transfer Value must be negative. Value: %d", value)
+	}
+	if transfer.dbTransfer.Value > math.MaxInt64 {
+		return fmt.Errorf("Value for transfer.dbTransfer exceeds MaxInt64: %d", transfer.dbTransfer.Value)
+	}
+	// Type conversion is safe here because of the MaxInt64 bounds check above
+	if value != -int64(transfer.dbTransfer.Value) {
+		return fmt.Errorf("Transfer is invalid: transfer.value %d must equal the inverse of transfer.dbTransfer.Value %d", value, transfer.dbTransfer.Value)
+	}
+	if targetChain != ce.emitterChainId {
+		return fmt.Errorf("Flow cancel transfer TargetChain %s does not match this chainEntry %s", targetChain, ce.emitterChainId)
+	}
+	// TODO: Verify the asset?
```

Comment:
What would you want to verify about the asset? Or can we get rid of this?

Reply:
My thinking was that since this method adds a Flow Cancel token, maybe it could validate that the asset being added in the `transfer` has the `flowCancels` bool property set to true. Thought it might be overkill but I suppose there's no harm?

Reply:
I pushed a new commit that removes this TODO and adds a comment that calling code has to do this verification. I also modified the unit tests to make sure that this verification is done.